In [58]:
#importing the required libraries
import numpy as np
import pandas as pd
df = pd.read_csv('C:\Python stuff\creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [18]:
#checking the number of rows and columns
df.shape

(284807, 31)

In [19]:
#summary of the dataset, along with the number of objects that do not have null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [20]:
#checking the rows that are duplicated
df[df.duplicated()]


Time        V1        V2        V3        V4        V5        V6  \
33          26.0 -0.529912  0.873892  1.347247  0.145457  0.414209  0.100223   
35          26.0 -0.535388  0.865268  1.351076  0.147575  0.433680  0.086983   
113         74.0  1.038370  0.127486  0.184456  1.109950  0.441699  0.945283   
114         74.0  1.038370  0.127486  0.184456  1.109950  0.441699  0.945283   
115         74.0  1.038370  0.127486  0.184456  1.109950  0.441699  0.945283   
...          ...       ...       ...       ...       ...       ...       ...   
282987  171288.0  1.912550 -0.455240 -1.750654  0.454324  2.089130  4.160019   
283483  171627.0 -1.464380  1.368119  0.815992 -0.601282 -0.689115 -0.487154   
283485  171627.0 -1.457978  1.378203  0.811515 -0.603760 -0.711883 -0.471672   
284191  172233.0 -2.667936  3.160505 -3.355984  1.007845 -0.377397 -0.109730   
284193  172233.0 -2.691642  3.123168 -3.339407  1.017018 -0.293095 -0.167054   

              V7        V8        V9  ...       V21       V22       V23  \
33      0.711206  0.176066 -0.286717  ...  0.046949  0.208105 -0.185548   
35      0.693039  0.179742 -0.285642  ...  0.049526  0.206537 -0.187108   
113    -0.036715  0.350995  0.118950  ...  0.102520  0.605089  0.023092   
114    -0.036715  0.350995  0.118950  ...  0.102520  0.605089  0.023092   
115    -0.036715  0.350995  0.118950  ...  0.102520  0.605089  0.023092   
...          ...       ...       ...  ...       ...       ...       ...   
282987 -0.881302  1.081750  1.022928  ... -0.524067 -1.337510  0.473943   
283483 -0.303778  0.884953  0.054065  ...  0.287217  0.947825 -0.218773   
283485 -0.282535  0.880654  0.052808  ...  0.284205  0.949659 -0.216949   
284191 -0.667233  2.309700 -1.639306  ...  0.391483  0.266536 -0.079853   
284193 -0.745886  2.325616 -1.634651  ...  0.402639  0.259746 -0.086606   

             V24       V25       V26       V27       V28  Amount  Class  
33      0.001031  0.098816 -0.552904 -0.073288  0.023307    6.14      0  
35      0.000753  0.098117 -0.553471 -0.078306  0.025427    1.77      0  
113    -0.626463  0.479120 -0.166937  0.081247  0.001192    1.18      0  
114    -0.626463  0.479120 -0.166937  0.081247  0.001192    1.18      0  
115    -0.626463  0.479120 -0.166937  0.081247  0.001192    1.18      0  
...          ...       ...       ...       ...       ...     ...    ...  
282987  0.616683 -0.283548 -1.084843  0.073133 -0.036020   11.99      0  
283483  0.082926  0.044127  0.639270  0.213565  0.119251    6.82      0  
283485  0.083250  0.044944  0.639933  0.219432  0.116772   11.93      0  
284191 -0.096395  0.086719 -0.451128 -1.183743 -0.222200   55.66      0  
284193 -0.097597  0.083693 -0.453584 -1.205466 -0.213020   36.74      0  

[1081 rows x 31 columns]

In [21]:
#deleting the duplicated rows in the original dataframe
df.drop_duplicates(inplace=True)

In [22]:
#checking the number of missing values in the dataset
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [23]:
#Finding out the mean, standard deviation and IQR va;ues
df.describe().T

count          mean           std         min           25%  \
Time    283726.0  94811.077600  47481.047891    0.000000  54204.750000   
V1      283726.0      0.005917      1.948026  -56.407510     -0.915951   
V2      283726.0     -0.004135      1.646703  -72.715728     -0.600321   
V3      283726.0      0.001613      1.508682  -48.325589     -0.889682   
V4      283726.0     -0.002966      1.414184   -5.683171     -0.850134   
V5      283726.0      0.001828      1.377008 -113.743307     -0.689830   
V6      283726.0     -0.001139      1.331931  -26.160506     -0.769031   
V7      283726.0      0.001801      1.227664  -43.557242     -0.552509   
V8      283726.0     -0.000854      1.179054  -73.216718     -0.208828   
V9      283726.0     -0.001596      1.095492  -13.434066     -0.644221   
V10     283726.0     -0.001441      1.076407  -24.588262     -0.535578   
V11     283726.0      0.000202      1.018720   -4.797473     -0.761649   
V12     283726.0     -0.000715      0.994674  -18.683715     -0.406198   
V13     283726.0      0.000603      0.995430   -5.791881     -0.647862   
V14     283726.0      0.000252      0.952215  -19.214325     -0.425732   
V15     283726.0      0.001043      0.914894   -4.498945     -0.581452   
V16     283726.0      0.001162      0.873696  -14.129855     -0.466860   
V17     283726.0      0.000170      0.842507  -25.162799     -0.483928   
V18     283726.0      0.001515      0.837378   -9.498746     -0.498014   
V19     283726.0     -0.000264      0.813379   -7.213527     -0.456289   
V20     283726.0      0.000187      0.769984  -54.497720     -0.211469   
V21     283726.0     -0.000371      0.723909  -34.830382     -0.228305   
V22     283726.0     -0.000015      0.724550  -10.933144     -0.542700   
V23     283726.0      0.000198      0.623702  -44.807735     -0.161703   
V24     283726.0      0.000214      0.605627   -2.836627     -0.354453   
V25     283726.0     -0.000232      0.521220  -10.295397     -0.317485   
V26     283726.0      0.000149      0.482053   -2.604551     -0.326763   
V27     283726.0      0.001763      0.395744  -22.565679     -0.070641   
V28     283726.0      0.000547      0.328027  -15.430084     -0.052818   
Amount  283726.0     88.472687    250.399437    0.000000      5.600000   
Class   283726.0      0.001667      0.040796    0.000000      0.000000   

                 50%            75%            max  
Time    84692.500000  139298.000000  172792.000000  
V1          0.020384       1.316068       2.454930  
V2          0.063949       0.800283      22.057729  
V3          0.179963       1.026960       9.382558  
V4         -0.022248       0.739647      16.875344  
V5         -0.053468       0.612218      34.801666  
V6         -0.275168       0.396792      73.301626  
V7          0.040859       0.570474     120.589494  
V8          0.021898       0.325704      20.007208  
V9         -0.052596       0.595977      15.594995  
V10        -0.093237       0.453619      23.745136  
V11        -0.032306       0.739579      12.018913  
V12         0.139072       0.616976       7.848392  
V13        -0.012927       0.663178       7.126883  
V14         0.050209       0.492336      10.526766  
V15         0.049299       0.650104       8.877742  
V16         0.067119       0.523512      17.315112  
V17        -0.065867       0.398972       9.253526  
V18        -0.002142       0.501956       5.041069  
V19         0.003367       0.458508       5.591971  
V20        -0.062353       0.133207      39.420904  
V21        -0.029441       0.186194      27.202839  
V22         0.006675       0.528245      10.503090  
V23        -0.011159       0.147748      22.528412  
V24         0.041016       0.439738       4.584549  
V25         0.016278       0.350667       7.519589  
V26        -0.052172       0.240261       3.517346  
V27         0.001479       0.091208      31.612198  
V28         0.011288       0.078276      33.847808  
Amount     22.000000      77.510000   25691.160000  

In [29]:
#Trying to identify the imbalance in the dataset
print("The number of non-fraudulent transactions" ,len(df[df['Class']==0]))
print("The number of fraudulent transactions" ,len(df[df['Class']==1]))


The number of non-fraudulent transactions 283253
The number of fraudulent transactions 473


In [59]:
#splitting the dataset into dependent and independent features
x = df.drop(['Class'],axis = 1)
x.head()
x.shape
y = df["Class"]
y.head()
y.shape


(284807,)

In [61]:
#performing logistic regression without balancing
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X_1, Y_1, test_size =0.20, random_state = 40)

In [62]:
#Performing Logistic Regression
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, Y_train)


C:\Users\harsh\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [31]:
pip install -U imbalanced-learn

Note: you may need to restart the kernel to use updated packages.


In [32]:
#performing undersampling on the above dataset
from imblearn import under_sampling
from imblearn.under_sampling import RandomUnderSampler 


In [42]:
#As shown the sampling size has been decreased from 283253 to 946
under_sampler = RandomUnderSampler()
X_1, Y_1 = under_sampler.fit_resample(x,y)
X_1.shape
Y_1.shape

(946,)

In [43]:
X_1.head()
Y_1.head()

0    0
1    0
2    0
3    0
4    0
Name: Class, dtype: int64

In [44]:
#splitting the data
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X_1, Y_1, test_size =0.20, random_state = 40)
print("Precision, Recall and F1scores\n ",classification_report(Y_test, y_predict))

In [63]:
#Performing Logistic Regression
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, Y_train)
print("The model score is ", model.score(X_test,Y_test))
print("The model coefficient for the prediction of each target is \n", model.coef_)
print("The model intercept is \n", model.intercept_)



The model score is  0.9410954793384053
The model coefficient for the prediction of each target is 
 [[-2.71085711e-05  7.78767171e-02 -5.14375963e-02 -4.71939607e-01
   6.55947003e-01  3.33623241e-01 -1.68759871e-01  5.60190350e-02
  -4.26046938e-01 -1.96574973e-01 -2.73197757e-01  7.01747463e-02
  -4.10660846e-01 -2.60422415e-01 -8.27181514e-01 -1.82236855e-01
  -2.46686627e-01 -2.44566906e-01  5.16770129e-02  4.35533106e-02
  -7.56406856e-02  1.05903601e-01  2.58162643e-01 -2.92839088e-02
   8.51374571e-03 -1.30203438e-01 -5.56797335e-02  2.71769797e-02
   1.90406956e-02  9.29173191e-04]]
The model intercept is 
 [-0.36634268]


C:\Users\harsh\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [69]:
model = LogisticRegression(solver='liblinear', random_state=40).fit(X_train, Y_train)
predict_val = model.predict(X_test)
print("The accuracy score is ",accuracy_score(Y_test, predict_val))
print("Precision, Recall and F1scores\n ",classification_report(Y_test, predict_val))

The accuracy score is  0.9344760021888404
Precision, Recall and F1scores
                precision    recall  f1-score   support

           0       0.90      0.98      0.94     56602
           1       0.97      0.89      0.93     56700

    accuracy                           0.93    113302
   macro avg       0.94      0.93      0.93    113302
weighted avg       0.94      0.93      0.93    113302



In [37]:
y_predict = model.predict(X_test)
y_predict

array([0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0], dtype=int64)

In [48]:
print("The model score is ", model.score(X_test,Y_test))
print("The model coefficient for the prediction of each target is \n", model.coef_)
print("The model intercept is \n", model.intercept_)


The model score is  0.9210526315789473
The model coefficient for the prediction of each target is 
 [[-2.65292673e-05 -1.25184081e-02  1.78618672e-01 -7.92506333e-01
   8.18083425e-01 -2.44596960e-03 -2.69170921e-01 -2.65025597e-01
   4.54430768e-02 -2.59726861e-01 -4.86448126e-01  4.01983837e-01
  -6.55452051e-01 -1.67610020e-02 -1.03656799e+00 -1.16006311e-01
  -3.01712931e-01 -5.59274581e-01 -8.58618838e-02  4.12036339e-02
  -5.71340872e-02  5.85108426e-02  6.70442158e-02 -8.29435778e-02
  -1.06656552e-02 -3.66673907e-02 -3.56014120e-02  3.93272391e-02
   1.32228031e-02  5.51615946e-04]]
The model intercept is 
 [-0.14756324]


In [50]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(Y_test, y_predict)
print("Confusion matrix is as follows ",cm)
print("The accuracy score is ",accuracy_score(Y_test, y_predict))
print("Precision, Recall and F1scores\n ",classification_report(Y_test, y_predict))

Confusion matrix is as follows  [[91  5]
 [11 83]]
The accuracy score is  0.9157894736842105
Precision, Recall and F1scores
                precision    recall  f1-score   support

           0       0.89      0.95      0.92        96
           1       0.94      0.88      0.91        94

    accuracy                           0.92       190
   macro avg       0.92      0.92      0.92       190
weighted avg       0.92      0.92      0.92       190



In [51]:
#Oversampling the data
from imblearn import over_sampling
from imblearn.over_sampling import RandomOverSampler 
over_sampler = RandomOverSampler()
X_1, Y_1 = over_sampler.fit_resample(x,y)
X_1.shape
Y_1.shape

(566506,)

In [ ]:
X_1.head()
Y_1.head()

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X_1, Y_1, test_size =0.20, random_state = 40)


In [52]:
#Performing Logistic Regression by oversampling
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, Y_train)
y_predict = model.predict(X_test)
y_predict


C:\Users\harsh\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


array([0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0], dtype=int64)

In [53]:
model.score(X_test,Y_test)

0.9157894736842105

In [54]:
model.coef_

array([[-3.02021535e-05, -7.44266866e-02, -1.26693817e-02,
        -4.98766518e-01,  8.15762575e-01,  3.86246474e-01,
        -2.00986089e-01,  1.85390573e-01, -7.01701624e-01,
         6.54962931e-03, -8.47359634e-02,  1.64595193e-01,
        -1.97745735e-01, -1.10994229e-01, -1.27275753e+00,
        -2.68464312e-01, -1.15653930e-01, -3.73066013e-01,
        -1.71925491e-03, -1.37544971e-02, -2.03533907e-01,
         9.01234765e-02,  4.29952028e-01, -2.55731181e-01,
        -2.94359033e-02, -1.64512362e-01, -1.81293906e-01,
         1.54282871e-01,  2.26562351e-02,  1.39858828e-03]])

In [55]:
model.intercept_

array([-0.50568219])

In [57]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
cm = confusion_matrix(Y_test, y_predict)
print("The confusion matric is as follows ",cm)
print("The accuracy score is ",accuracy_score(Y_test, y_predict))
print("The precision, recall and f-1scores are\n",classification_report(Y_test, y_predict))

The confusion matric is as follows  [[90  6]
 [10 84]]
The accuracy score is  0.9157894736842105
The precision, recall and f-1scores are
               precision    recall  f1-score   support

           0       0.90      0.94      0.92        96
           1       0.93      0.89      0.91        94

    accuracy                           0.92       190
   macro avg       0.92      0.92      0.92       190
weighted avg       0.92      0.92      0.92       190



Conclusion:
* In the above code, we performed logistic Regression without any balancing techniques and with undersampling and oversampling.
* As the data is highly imbalanced, in order to understand the difference balancing can make, we performed the regression without any kind of class balance, through this, we found out that the accuracy is unsually high.
* When we balanced the weights through both, undersampling and oversampling, the accuracy came out to be 93 and 91% respectively, indicating that if class weights are imbalanced, it can affect the accuracy of the model drastically.